File: adapters/hypermeow/joiner.go
package wawire

import (
	"sync"

	"github.com/oxidezap/wa-wire/adapters/hypermeow/wire"
	waBinary "go.mau.fi/whatsmeow/binary"
)

// DefaultLookahead is how many later stanzas a pending message tolerates
// before it is emitted with whatever it has.
//
// Sized for the widest real fan-out rather than tuned: a message's payloads
// all arrive within its own processing, so a handful of intervening stanzas
// already means they are not coming.
// DefaultLookahead must exceed the engine's handler queue, and does.
//
// The two hooks sit on opposite sides of that queue: the raw-node hook runs on
// the receive goroutine, and a plaintext arrives from `handlerQueueLoop`, which
// buffers 256 nodes. So a message can legitimately lag by the whole queue
// before its payload appears, and a lookahead under that would give up on
// traffic that was about to arrive — recording `Unobserved` for a stanza that
// decrypted perfectly well.
//
// Twice the queue, so the margin does not depend on reading the engine's
// constant exactly right. The Rust and TypeScript adapters use 64 because
// their engines hand both halves over on one path and no queue sits between.
const DefaultLookahead = 512

// Joiner holds a stanza until the plaintexts decrypted out of it catch up.
//
// The two arrive separately and cannot be made to arrive together. The raw-node
// hook runs inside the Noise frame callback, necessarily before Signal has run;
// a plaintext exists only afterwards.
//
// # Knowing when to stop waiting
//
// The stanza says how many `<enc>` children it has, so the common case — every
// one decrypts — closes by counting, with no clock: the last payload completes
// the table and the envelope goes out at once.
//
// What has no signal is an `<enc>` that will never produce a payload. So
// something has to give up, and giving up is measured in **stanzas rather than
// milliseconds**: the receive path is ordered, so a stanza whose payloads have
// not arrived after DefaultLookahead later ones is a stanza whose payloads are
// not coming. A count is also the same on every machine, which a duration is
// not, and this output is compared against other engines'.
//
// The Rust and TypeScript adapters reach the same conclusions from the same
// constraints (wa-wire DESIGN D-052, D-053, D-055). The three are deliberately
// alike: an adapter that decided differently would produce recordings that
// differ for reasons the engines are not responsible for.
//
// # What this adapter does not have to do
//
// Address a payload to its node. The other two engines report which `<enc>` of
// a stanza decrypted, counting `<enc>` nodes, and an adapter has to work out
// which child that is — ambiguous the moment a stanza carries anything else,
// and unresolvable for a fan-out `<message>`, where the copies for this device
// are numbered apart from the direct children.
//
// hypermeow reports the child index directly, so the path is that index and
// nothing is inferred. The hook was written against this need.
type Joiner struct {
	mu sync.Mutex
	// Every stanza seen, in the order the receive goroutine saw them.
	//
	// A queue rather than a set of pending ones, because emitting an unheld
	// stanza the moment it arrives reorders it ahead of a held one that came
	// first — and a recording compared position by position would call that a
	// divergence. The front of the queue drains whenever the front is ready,
	// so what leaves is what arrived, in that order.
	queue   []*slot
	byID    map[string]*slot
	pending int

	lookahead int

	// Held across the sink call and nothing else.
	//
	// The two hooks run on different goroutines — the raw-node one on the
	// receive path, the plaintext one on the engine's handler queue — so
	// without this two deliveries can be in `Accept` at once. A sink is a
	// consumer's code and is entitled not to be reentrant.
	//
	// Separate from `mu` so a slow sink stalls deliveries rather than the
	// engine's receive path.
	emitMu sync.Mutex
	sink   Sink
}

// slot is one stanza's place in the queue.
type slot struct {
	// id is the message id, empty for a stanza that is not held.
	id    string
	frame []byte
	// The child indices of this stanza's `<enc>` nodes, in document order.
	//
	// The indices themselves rather than a count: an `<enc>` is not
	// necessarily the first child, and a stanza whose `<enc>` nodes sit at
	// positions two and three has nothing at zero and one. Keeping the real
	// positions is also what lets an `<enc>` that produced nothing be reported
	// at the path it actually occupies.
	encIndices []int
	plaintexts map[int]wire.Plaintext
	// How many stanzas have arrived since this one.
	age int
	// Whether this slot is finished and only waiting for its turn.
	done bool
}

func (s *slot) ready() bool {
	return s.done || len(s.plaintexts) >= len(s.encIndices)
}

// Sink receives envelopes as they complete.
//
// Called with the joiner's lock released, so a slow sink does not stall the
// engine's receive path — but the calls are ordered, and a sink that blocks
// still delays the envelopes behind it.
type Sink interface {
	Accept(envelope wire.Envelope)
}

// SinkFunc adapts a function to a Sink.
type SinkFunc func(envelope wire.Envelope)

// Accept calls the function.
func (f SinkFunc) Accept(envelope wire.Envelope) { f(envelope) }

// NewJoiner returns a joiner emitting to sink.
func NewJoiner(sink Sink) *Joiner {
	return &Joiner{
		byID:      make(map[string]*slot),
		lookahead: DefaultLookahead,
		sink:      sink,
	}
}

// WithLookahead sets how many later stanzas a pending one tolerates. For tests
// and for a host that knows its own traffic; the default suits this engine's
// handler queue.
func (j *Joiner) WithLookahead(stanzas int) *Joiner {
	j.lookahead = stanzas
	return j
}

// AcceptFrame takes a decoded stanza and its bytes.
//
// A stanza with no `<enc>` children is finished on arrival; one with them waits
// for its payloads. Either way it takes its place in the queue, and the queue
// drains in order — so an ack behind a held message stays behind it, which is
// where the wire put it.
//
// The frame is copied. The engine documents its buffer as valid only for the
// duration of the call — an uncompressed frame is a window into the transport's
// own read buffer — and queueing means outliving the call by construction.
func (j *Joiner) AcceptFrame(node *waBinary.Node, frame []byte) {
	encIndices := encChildIndices(node)
	id, _ := node.Attrs["id"].(string)
	// A stanza carrying `<enc>` children and no id is not one this can hold: it
	// would never be matched to its payloads. Finished on arrival, as L0-wire.
	holdable := len(encIndices) > 0 && id != ""

	j.mu.Lock()
	// Every stanza ages the ones waiting, including the ones finished on
	// arrival: the lookahead counts *later stanzas*, and an ack going past is
	// one. A receive path carrying nothing but acks would otherwise hold a
	// message for ever.
	j.ageLocked()

	entry := &slot{frame: cloneBytes(frame)}
	if holdable {
		entry.id = id
		entry.encIndices = encIndices
		entry.plaintexts = make(map[int]wire.Plaintext, len(encIndices))
		// A repeated id — a retry arriving before the first one finished —
		// leaves the earlier stanza in the queue and unreachable by id rather
		// than replacing it. It was a real stanza and this adapter promises to
		// report every one; what it loses is the chance of a payload, which
		// the earlier one gives up as `Unobserved`.
		if previous, held := j.byID[id]; held {
			previous.done = true
		}
		j.byID[id] = entry
		j.pending++
	} else {
		entry.done = true
	}
	j.queue = append(j.queue, entry)
	ready := j.drainLocked()
	j.mu.Unlock()

	j.emitAll(ready)
}

// AcceptPlaintext takes one decrypted payload and completes its stanza if this
// was the last one outstanding.
func (j *Joiner) AcceptPlaintext(messageID string, childIndex int, plaintext []byte) {
	j.mu.Lock()
	entry, held := j.byID[messageID]
	if !held || entry.done {
		// No stanza to attach it to: one already given up on, or a frame never
		// seen. Inventing somewhere to put it would be worse than losing it.
		j.mu.Unlock()
		return
	}
	if !entry.expects(childIndex) {
		// A payload for a child this stanza has no `<enc>` at. The engine and
		// the frame disagree about the stanza, and counting it would close the
		// stanza early and drop the payload that was actually coming.
		j.mu.Unlock()
		return
	}
	entry.plaintexts[childIndex] = wire.Plaintext{
		Path:    []uint16{uint16(childIndex)},
		Status:  wire.StatusOk,
		Payload: cloneBytes(plaintext),
	}
	ready := j.drainLocked()
	j.mu.Unlock()

	j.emitAll(ready)
}

// Flush emits every stanza still queued, with whatever it has.
//
// For shutdown: a frame still waiting for a payload that will now never arrive
// is better emitted unobserved than lost, since the stanza was real either way.
func (j *Joiner) Flush() {
	j.mu.Lock()
	for _, entry := range j.queue {
		entry.done = true
	}
	ready := j.drainLocked()
	j.mu.Unlock()

	j.emitAll(ready)
}

// Pending reports how many stanzas are waiting on payloads.
func (j *Joiner) Pending() int {
	j.mu.Lock()
	defer j.mu.Unlock()
	return j.pending
}

// Queued reports how many stanzas are in the queue, waiting or merely behind
// one that is.
func (j *Joiner) Queued() int {
	j.mu.Lock()
	defer j.mu.Unlock()
	return len(j.queue)
}

func (j *Joiner) ageLocked() {
	for _, entry := range j.queue {
		if entry.done {
			continue
		}
		entry.age++
		if entry.age > j.lookahead {
			entry.done = true
		}
	}
}

// drainLocked takes the finished stanzas off the front of the queue.
//
// The front, and only the front: a finished stanza behind an unfinished one
// waits, because the unfinished one arrived first and that is the order the
// wire had.
func (j *Joiner) drainLocked() []wire.Envelope {
	var ready []wire.Envelope
	at := 0
	for ; at < len(j.queue); at++ {
		entry := j.queue[at]
		if !entry.ready() {
			break
		}
		ready = append(ready, entry.envelope())
		if entry.id != "" {
			if current, held := j.byID[entry.id]; held && current == entry {
				delete(j.byID, entry.id)
			}
			j.pending--
		}
	}
	j.queue = j.queue[at:]
	return ready
}

// emitAll delivers in order, and one at a time.
//
// The lock is this adapter's own rather than the engine's: two hooks on two
// goroutines both reach here, and a sink is a consumer's code that is entitled
// not to be reentrant.
func (j *Joiner) emitAll(envelopes []wire.Envelope) {
	if len(envelopes) == 0 || j.sink == nil {
		return
	}
	j.emitMu.Lock()
	defer j.emitMu.Unlock()
	for _, envelope := range envelopes {
		j.sink.Accept(envelope)
	}
}

// expects reports whether this stanza has an `<enc>` at that child index.
func (s *slot) expects(childIndex int) bool {
	for _, index := range s.encIndices {
		if index == childIndex {
			return true
		}
	}
	return false
}

// envelope builds what this stanza has.
//
// An `<enc>` that produced nothing is reported as `Unobserved` rather than
// omitted: a table missing an entry says the node was not there, and it was.
// `Unobserved` claims only what this adapter can see — that no payload arrived
// — and not why, which it cannot.
func (s *slot) envelope() wire.Envelope {
	if len(s.encIndices) == 0 {
		return wire.Envelope{Frame: s.frame}
	}
	entries := make([]wire.Plaintext, 0, len(s.encIndices))
	for _, index := range s.encIndices {
		if entry, seen := s.plaintexts[index]; seen {
			entries = append(entries, entry)
			continue
		}
		entries = append(entries, wire.Plaintext{
			Path:   []uint16{uint16(index)},
			Status: wire.StatusUnobserved,
		})
	}
	return wire.Envelope{Frame: s.frame, Plaintexts: entries}
}

// encChildIndices are the positions of a stanza's `<enc>` children among all
// of its children — the same numbering the engine reports a payload under.
func encChildIndices(node *waBinary.Node) []int {
	var indices []int
	for index, child := range node.GetChildren() {
		if child.Tag == "enc" {
			indices = append(indices, index)
		}
	}
	return indices
}

func cloneBytes(source []byte) []byte {
	if source == nil {
		return nil
	}
	out := make([]byte, len(source))
	copy(out, source)
	return out
}

File: adapters/baileys/src/joiner.ts
/**
 * Joining a stanza to the plaintexts decrypted out of it.
 *
 * The two arrive separately and cannot be made to arrive together. The frame
 * hook runs where a stanza is decoded, which is necessarily before Signal has
 * run; a plaintext exists only afterwards.
 *
 * # Knowing when to stop waiting
 *
 * The stanza says how many `<enc>` children it has, so the common case — every
 * one decrypts — closes by counting, with no clock: the last payload completes
 * the table and the envelope goes at once.
 *
 * What has no signal is an `<enc>` that will never produce a payload. So
 * something has to give up, and giving up is measured in **stanzas rather than
 * milliseconds**: the receive path is ordered, and a count is the same on every
 * machine, which a duration is not — and this output is compared against other
 * engines'.
 *
 * # Stanzas leave in the order they arrived
 *
 * A stanza waiting on payloads is not a licence to reorder the ones behind it.
 * Emitting an ack the moment it arrives would put it ahead of a message that
 * came first, and a recording compared position by position would report that
 * as a divergence in whichever engine happened to be slower.
 *
 * The `zapo` and `whatsapp-rust` adapters emit unheld stanzas immediately and
 * do reorder. They were written before this was understood; the Go adapter and
 * this one queue. Worth reconciling — three adapters ordering differently is a
 * difference the engines are not responsible for.
 *
 * # What this adapter does not have to do
 *
 * Work out which node a payload belongs to. Baileys reports the child index
 * directly, so the path is that number and nothing is inferred.
 */

import { Direction, FrameOrigin, PlaintextStatus, type Plaintext, type Stanza } from '@oxidezap/wa-wire-ts'

/** A decoded stanza, as much of one as this adapter reads. */
export interface Node {
    readonly tag: string
    readonly attrs: Record<string, string | undefined>
    readonly content?: unknown
}

/**
 * How many later stanzas a pending message tolerates.
 *
 * Sized for the widest real fan-out rather than tuned: a message's payloads all
 * arrive within its own processing, so a handful of intervening stanzas already
 * means they are not coming.
 */
export const DEFAULT_LOOKAHEAD = 64

/** One decrypted payload, as the engine reports it. */
export interface DecryptedEnc {
    /** The stanza id this belongs to. */
    readonly messageId: string
    /** The position of the `<enc>` among the stanza's children. */
    readonly childIndex: number
    /** The plaintext. */
    readonly plaintext: Uint8Array
}

/** Receives envelopes as they complete, in the order their stanzas arrived. */
export type StanzaSink = (stanza: Stanza) => void

interface Slot {
    /** Empty for a stanza that is not held. */
    id: string
    frame: Uint8Array
    /** The child indices of this stanza's `<enc>` nodes, in document order. */
    encIndices: number[]
    plaintexts: Map<number, Plaintext>
    age: number
    done: boolean
}

const ready = (slot: Slot) => slot.done || slot.plaintexts.size >= slot.encIndices.length

export class PlaintextJoiner {
    /** Every stanza seen, in the order the receive path saw them. */
    private queue: Slot[] = []
    private byId = new Map<string, Slot>()

    constructor(private readonly lookahead: number = DEFAULT_LOOKAHEAD) {}

    /**
     * Take a decoded stanza and the bytes it came from.
     *
     * The frame is copied: the engine's buffer is a view into what it read, and
     * queueing means outliving the call.
     */
    public acceptFrame(node: Node, frame: Uint8Array, sink: StanzaSink): void {
        const encIndices = encChildIndices(node)
        const id = node.attrs.id ?? ''
        // A stanza carrying `<enc>` children and no id is not one this can
        // hold: it would never be matched to its payloads. Finished on arrival.
        const holdable = encIndices.length > 0 && id !== ''

        // Every stanza ages the ones waiting, including the ones finished on
        // arrival: the lookahead counts *later stanzas*, and an ack going past
        // is one. A receive path carrying nothing but acks would otherwise hold
        // a message for ever.
        this.age()

        const slot: Slot = {
            id: holdable ? id : '',
            frame: Uint8Array.prototype.slice.call(frame),
            encIndices,
            plaintexts: new Map(),
            age: 0,
            done: !holdable
        }
        if (holdable) {
            // A repeated id — a retry arriving before the first finished —
            // leaves the earlier stanza in the queue rather than replacing it.
            // It was a real stanza and this adapter reports every one; what it
            // gives up is the chance of a payload.
            const previous = this.byId.get(id)
            if (previous) {
                previous.done = true
            }

            this.byId.set(id, slot)
        }

        this.queue.push(slot)
        this.drain(sink)
    }

    /** Take one decrypted payload, completing its stanza if it was the last. */
    public acceptPlaintext(decrypted: DecryptedEnc, sink: StanzaSink): void {
        const slot = this.byId.get(decrypted.messageId)
        if (!slot || slot.done) {
            // No stanza to attach it to: one already given up on, or a frame
            // never seen. Inventing somewhere to put it would be worse.
            return
        }

        if (!slot.encIndices.includes(decrypted.childIndex)) {
            // A payload for a child this stanza has no `<enc>` at. The engine
            // and the frame disagree; counting it would close the stanza early
            // and drop the payload that was actually coming.
            return
        }

        slot.plaintexts.set(decrypted.childIndex, {
            path: [decrypted.childIndex],
            status: PlaintextStatus.Ok,
            payload: Uint8Array.prototype.slice.call(decrypted.plaintext)
        })
        this.drain(sink)
    }

    /**
     * Emit everything still queued, with whatever it has.
     *
     * For shutdown: a frame still waiting for a payload that will never arrive
     * is better emitted unobserved than lost, the stanza having been real.
     */
    public flush(sink: StanzaSink): void {
        for (const slot of this.queue) {
            slot.done = true
        }

        this.drain(sink)
    }

    /** How many stanzas are waiting on payloads. */
    public get pending(): number {
        return this.queue.filter(slot => !ready(slot)).length
    }

    /** How many are queued, waiting or merely behind one that is. */
    public get queued(): number {
        return this.queue.length
    }

    private age(): void {
        for (const slot of this.queue) {
            if (slot.done) {
                continue
            }

            slot.age += 1
            if (slot.age > this.lookahead) {
                slot.done = true
            }
        }
    }

    /**
     * Take the finished stanzas off the front.
     *
     * The front, and only the front: a finished stanza behind an unfinished one
     * waits, because the unfinished one arrived first and that is the order the
     * wire had.
     */
    private drain(sink: StanzaSink): void {
        let at = 0
        for (; at < this.queue.length; at++) {
            const slot = this.queue[at]!
            if (!ready(slot)) {
                break
            }

            if (slot.id !== '' && this.byId.get(slot.id) === slot) {
                this.byId.delete(slot.id)
            }

            sink(envelopeOf(slot))
        }

        this.queue = this.queue.slice(at)
    }
}

/**
 * What this stanza has.
 *
 * An `<enc>` that produced nothing is reported as `Unobserved` rather than
 * omitted: a table missing an entry says the node was not there, and it was.
 * `Unobserved` claims only what this adapter can see — that no payload arrived
 * — and not why, which it cannot.
 */
const envelopeOf = (slot: Slot): Stanza => {
    if (slot.encIndices.length === 0) {
        return { direction: Direction.Inbound, frameOrigin: FrameOrigin.Original, frame: slot.frame }
    }

    const plaintexts = slot.encIndices.map(
        index =>
            slot.plaintexts.get(index) ?? {
                path: [index],
                status: PlaintextStatus.Unobserved,
                payload: new Uint8Array(0)
            }
    )

    return {
        direction: Direction.Inbound,
        frameOrigin: FrameOrigin.Original,
        frame: slot.frame,
        plaintexts
    }
}

/**
 * The positions of a stanza's `<enc>` children among all of its children.
 *
 * The same numbering the engine reports a payload under — an `<enc>` is not
 * necessarily the first child, and counting `<enc>` nodes instead would address
 * a plaintext to the wrong node.
 */
const encChildIndices = (node: Node): number[] => {
    if (!Array.isArray(node.content)) {
        return []
    }

    const indices: number[] = []
    for (const [index, child] of node.content.entries()) {
        if (typeof child === 'object' && child !== null && (child as Node).tag === 'enc') {
            indices.push(index)
        }
    }

    return indices
}
